internal/controllers/gateway/httproute_controller.go

```diff
@@ -752,3 +761,39 @@ func (r *HTTPRouteReconciler) getHTTPRouteRuleReason(ctx context.Context, httpRo
 func (r *HTTPRouteReconciler) SetLogger(l logr.Logger) {
 	r.Log = l
 }
+
+// ensureNoStaleParentStatus removes any status references to Gateways that are no longer in the HTTPRoute's parentRefs
+// and returns true if any status was removed.
+func ensureNoStaleParentStatus(httproute *gatewayapi.HTTPRoute) (wasAnyStatusRemoved bool) {
+	// Create a map of currently defined parentRefs for fast lookup.
+	currentlyDefinedParentRefs := make(map[string]struct{})
+	for _, parentRef := range httproute.Spec.ParentRefs {
+		currentlyDefinedParentRefs[parentReferenceKey(httproute.Namespace, parentRef)] = struct{}{}
+	}
+
+	for parentIdx, parentStatus := range httproute.Status.Parents {
+		// Don't touch statuses from other controllers.
+		if parentStatus.ControllerName != GetControllerName() {
+			continue
+		}
+		// Remove the status if the parentRef is no longer defined.
+		if _, ok := currentlyDefinedParentRefs[parentReferenceKey(httproute.Namespace, parentStatus.ParentRef)]; !ok {
+			httproute.Status.Parents = slices.Delete(httproute.Status.Parents, parentIdx, parentIdx+1)
+			wasAnyStatusRemoved = true
+		}
+	}
+	return wasAnyStatusRemoved
+}
+
+// parentReferenceKey returns a string key for a parentRef of a route. It can be used for indexing a map.
+func parentReferenceKey(routeNamespace string, parentRef gatewayapi.ParentReference) string {
+	namespace := routeNamespace
+	if parentRef.Namespace != nil {
+		namespace = string(*parentRef.Namespace)
+	}
+	sectionName := ""
+	if parentRef.SectionName != nil {
+		sectionName = string(*parentRef.SectionName)
+	}
+	return fmt.Sprintf("%s/%s/%s", namespace, parentRef.Name, sectionName)
```

Comment:
I think we need to include at least `port` in here also. SectionName is optional; AFAIK you can create match criteria that's just "the one Listener on port 80 that accepts HTTPRoutes" and not include its name.

Group and Kind technically also differentiate ParentRefs: AFAIK you can technically have a Gateway ref and a Service (or some vendor-specific kind) ref that have otherwise identical fields. This probably isn't likely to matter in practice and _definitely_ doesn't for us, since we don't support other kinds of parent.

Reply:
Yeah, you're right. Thanks for catching this. I've made it use port as well and covered the function with UTs: 
20a7dfa

I think we can safely ignore Group and Kind as that's validated in our admission webhook explicitly: https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/admission/validation/gateway/httproute.go#L79